controller/orphan_controller.go

```diff
@@ -269,48 +303,200 @@ func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
+			return true, nil
+		}
+		err = oc.deleteOrphanedReplicaDataStore(orphan)
+		if err == nil || (err != nil && apierrors.IsNotFound(err)) {
+			isCleanupComplete = true
+			err = nil
+		}
+	default:
+		return false, fmt.Errorf("unknown orphan type %v to clean up orphaned resource for %v", orphan.Spec.Type, orphan.Name)
 	}
 
-	switch orphan.Spec.Type {
-	case longhorn.OrphanTypeReplica:
-		err = oc.deleteOrphanedReplica(orphan)
+	return isCleanupComplete, err
+}
+
+func (oc *OrphanController) cleanupOrphanedEngineInstance(orphan *longhorn.Orphan) (isCleanupComplete bool, err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "cannot cleanup orphan engine instance %v", orphan.Name)
+		}
+	}()
+
+	instance, imName, err := oc.extractOrphanedInstanceInfo(orphan)
+	if err != nil {
+		return false, err
+	}
+
+	var spec *longhorn.InstanceSpec
+	if engineCR, err := oc.ds.GetEngineRO(instance); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		spec = nil
+	} else {
+		spec = &engineCR.Spec.InstanceSpec
+	}
+	return oc.cleanupOrphanedInstance(orphan, instance, imName, longhorn.InstanceManagerTypeEngine, orphan.Spec.DataEngine, spec)
+}
+
+func (oc *OrphanController) cleanupOrphanedReplicaInstance(orphan *longhorn.Orphan) (isCleanupComplete bool, err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "cannot cleanup orphan replica instance %v", orphan.Name)
+		}
+	}()
+
+	instance, imName, err := oc.extractOrphanedInstanceInfo(orphan)
+	if err != nil {
+		return false, err
+	}
+
+	var spec *longhorn.InstanceSpec
+	if replicaCR, err := oc.ds.GetReplicaRO(instance); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		spec = nil
+	} else {
+		spec = &replicaCR.Spec.InstanceSpec
+	}
+	return oc.cleanupOrphanedInstance(orphan, instance, imName, longhorn.InstanceManagerTypeReplica, orphan.Spec.DataEngine, spec)
+}
+
+func (oc *OrphanController) cleanupOrphanedInstance(orphan *longhorn.Orphan, instance, imName string, imType longhorn.InstanceManagerType, deType longhorn.DataEngineType, instanceCRSpec *longhorn.InstanceSpec) (isCleanupComplete bool, err error) {
+	if instanceCRSpec != nil && instanceCRSpec.NodeID == oc.controllerID {
+		// Instance is scheduled back to current node and is no longer considered orphaned.
+		// Finalize the orphan CR directly without terminating the instance process.
+		return true, nil
+	}
+
+	im, err := oc.ds.GetRunningInstanceManagerByNodeRO(oc.controllerID, deType)
+	if err != nil {
+		oc.logger.WithError(err).Infof("No running instance manager for node %v for deleting orphan instance %v", oc.controllerID, orphan.Name)
+		return true, nil
+	}
+	if im.Name != imName {
+		oc.logger.WithField("instanceManager", imName).Infof("Orphan instance %v is not managed by current running instance manager %v", orphan.Name, im.Name)
+		return true, nil
+	}
```

Comment:
nit: wrap the codes into a helper function like `isCleanupOrphanedInstanceRequired`?

Reply:
im is needed to create instance manager client later. Will extract to something like `getRunningInstanceManagerClientForOrphan`, and skip deletion if there's no available client without error.

Reply:
Okay. Not easy to abstract this part. Let's keep it as it is. Thank you.